Test bundle builds job before push

FoamyGuy opened the pull request:
Resolves: #377

This adds a new job to the bundle cron action which builds both the community and adafruit bundles. 

The `update-bundles` job has had `needs: test-bundle-builds` added to it so that it depends upon the prior test builds job. If that test build fails then update-bundles will be skipped. 

I tested this to the best of my abilities in a test repo. There are a few failed runs visible here: https://github.com/FoamyGuy/adabot_actions_tests/actions which validated the way that `needs` will skip the job if it's dependency failed. For testing I modified the task to add an erroneous extra file into one of the libraries similarly to what caused the recent build failures.

That being said, I'm not sure if there is a great way to test it within the cron task context without just merging it and checking in after the next time it runs. And if there is a desire to test the failure case we'd need to temporarily modify the test build action to make one or both of the builds fail. If anyone does know of a better way to test I'm happy to give it a try.

tannewt reviewed (changes requested):
I think you want to do a test build *after* the update happens but before it is commit and then released. This prevents a second bad build but not the first.

FoamyGuy commented:
I am not sure that I understand. Do you mean handle this logic in python code after this line: https://github.com/adafruit/adabot/blob/main/adabot/circuitpython_bundle.py#L563 but before the following ones that commit and push it, rather than having the logic in an actions job only?

My understanding of the current implementation is that it won't allow a first bad build because the task fetches all submodules recursively here: https://github.com/FoamyGuy/adabot/blob/test_build_before_push/.github/workflows/bundle_cron.yml#L41 before doing the test build. And then if the test build fails the entire `update-bundles` job will be skipped due to the added `needs: test-bundle-builds`. Which leads to `adabot.circuitpython_bundle` (https://github.com/FoamyGuy/adabot/blob/test_build_before_push/.github/workflows/bundle_cron.yml#L87) not being executed so no further building, nor committing or pushing will occur since they are inside of that.

I am still happy to change it up if there is a better way, or if I've misunderstood how the actions will behave. Maybe it's just nicer to keep it all in the python code rather than split between python and workflows.

FoamyGuy commented:
The latest commits move the test build logic into the python code after the update and before the commit / push.

A few things about how it works now:

- I changed it to re-raise the RuntimeError that was previously getting caught and consumed. That change is in order to make the actions task fail when there is a failure to build any bundle. Without this change the failed bundle wouldn't be commited or pushed, but the actions task would still succeed. I don't see any other code path that raises the Runtime error, but I only looked briefly and I'm not sure if this change could cause other issues during different operations that happen inside the try block.
- Because the Adafruit bundle comes first in the list it means that if the Adafruit bundle builds successfully then it will get committed and pushed even if it turns out that the community bundle build fails.
- I attempted to test as best as I can over in the branch: https://github.com/adafruit/adabot/tree/refs/heads/actions_bundlebuild_test. In that branch I removed the cron task and set it up for on push so that I could get it to run "manually". When both bundles build successfully the task succeeds. When a bundle fails to build the task fails, here is one such failure log: https://github.com/adafruit/adabot/actions/runs/11194745798 I made it fail by manipulating one of the libraries after the bundle updates and before the test build.

tannewt commented on adabot/circuitpython_bundle.py:
```diff
@@ -419,6 +420,26 @@ def add_contributors(master_list, additions):
+def test_bundle_build(bundle_dir):
+    """
+    Attempts to build the bundle at the given location.
+    Returns exit code 0 if success.
+    Returns exit code >0 if failed to build.
```
I don't see a return value.

```suggestion
```

FoamyGuy replied:
@tannewt I'm not certain of the proper terminology, "Returns" is probably not correct. Maybe "Raises"? Basically these comments are talking about the sys.exit(2) from here: https://github.com/adafruit/circuitpython-build-tools/blob/1f3c5bd0a250dc162306a522c3ee5cd2c117716b/circuitpython_build_tools/scripts/build_bundles.py#L208

test_build_bundle function will allow that SystemExit (2) value to bubble upward which then gets caught down at 589. 


Alternatively I could move the try / catch SystemExit(2)  to inside of test_build_bundle() and then actually make it return the integer value so that the current documentation wording will be accurate.

jepler replied:
"Raises SystemExit with a nonzero value if failed to build"?

FoamyGuy replied:
re-worded with the latest commit. Thank you @jepler!

tannewt reviewed (changes requested):
Thanks! This looks more correct to me. One doc change. Good otherwise.

jepler commented:
Python documentation (pydoc) for `sys.exit` says the numeric argument is used as the "system exit status", and the docs.python.org documentation for SystemExit uses the same phrase. So if you're looking for an official-ese word for this, that's what I'd go with.

tannewt reviewed (approved):
This is better! Thank you!
